src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.cc

```diff
@@ -0,0 +1,140 @@
+
+#include <string.h>
+
+#include "crimson/common/log.h"
+
+#include "crimson/os/seastore/seastore_types.h"
+#include "crimson/os/seastore/omap_manager/btree/btree_omap_manager.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore::omap_manager {
+
+BtreeOMapManager::BtreeOMapManager(
+  TransactionManager &tm)
+  : tm(tm) {}
+
+BtreeOMapManager::initialize_omap_ret
+BtreeOMapManager::initialize_omap(Transaction &t)
+{
+
+  logger().debug("{}", __func__);
+  return tm.alloc_extent<OMapLeafNode>(t, L_ADDR_MIN, OMAP_BLOCK_SIZE)
+    .safe_then([this](auto&& root_extent) {
+      root_extent->set_size(0);
+      omap_node_meta_t meta{1};
+      root_extent->set_meta(meta);
+      omap_root_t omap_root = omap_root_t(1, root_extent->get_laddr());
+      return initialize_omap_ertr::make_ready_future<omap_root_t>(omap_root);
+  });
+}
+
+BtreeOMapManager::get_root_ret
+BtreeOMapManager::get_omap_root(omap_root_t &omap_root, Transaction &t)
+{
+  assert(omap_root.omap_root_laddr != L_ADDR_NULL);
+  laddr_t laddr = omap_root.omap_root_laddr;
+  return omap_load_extent(get_omap_context(omap_root, t), laddr, omap_root.depth);
+}
+
+BtreeOMapManager::omap_get_value_ret
+BtreeOMapManager::omap_get_value(omap_root_t &omap_root, Transaction &t,
+                                 const std::string &key)
+{
+  logger().debug("{}: {}", __func__, key);
+  return get_omap_root(omap_root, t).safe_then([this, &omap_root, &t, &key](auto&& extent) {
+    return extent->get_value(get_omap_context(omap_root, t), key);
+  }).safe_then([](auto &&e) {
+    logger().debug("{}: {} -> {}", __func__, e.first, e.second);
+    return omap_get_value_ret(
+        omap_get_value_ertr::ready_future_marker{},
+        std::move(e));
+  });
+
+}
+
+BtreeOMapManager::omap_set_key_ret
+BtreeOMapManager::omap_set_key(omap_root_t &omap_root, Transaction &t,
+                             const std::string &key, const std::string &value)
+{
+  logger().debug("{}: {} -> {}", __func__, key, value);
+  return get_omap_root(omap_root, t).safe_then([this, &omap_root, &t, &key, &value](auto &&root) {
+    return root->insert(get_omap_context(omap_root, t), key, value);
+  }).safe_then([](auto ret) {
+      return omap_set_key_ret(
+        omap_set_key_ertr::ready_future_marker{},
+        omap_mutation_result_t(ret.status));
+  });
+
+}
+
+BtreeOMapManager::omap_rm_key_ret
+BtreeOMapManager::omap_rm_key(omap_root_t &omap_root, Transaction &t, const std::string &key)
+{
+  logger().debug("{}: {}", __func__, key);
+  return get_omap_root(omap_root, t).safe_then([this, &omap_root, &t, &key](auto extent) {
+    return extent->rm_key(get_omap_context(omap_root, t), key);
+  }).safe_then([](auto ret) {
+    return omap_rm_key_ret(
+      omap_rm_key_ertr::ready_future_marker{},
+      omap_mutation_result_t(ret.status));
```

Comment:
Same here, this bit should handle the final merge/split by updating the omap_root.